ziti/cmd/ziti/cmd/edge_controller/login.go

```diff
@@ -80,70 +81,199 @@ func newLoginCmd(f cmdutil.Factory, out io.Writer, errOut io.Writer) *cobra.Comm
 			return err
 		}
+
+		if !jsonParsed.ExistsP("data.token") {
+			return fmt.Errorf("no session token returned from login request to %v. Received: %v", host, jsonParsed.String())
+		}
+
+		var ok bool
+		o.Token, ok = jsonParsed.Path("data.token").Data().(string)
+
+		if !ok {
+			return fmt.Errorf("session token returned from login request to %v is not in the expected format. Received: %v", host, jsonParsed.String())
+		}
+
+		if !o.OutputJSONResponse {
+			fmt.Printf("Token: %v\n", o.Token)
+		}
 	}
-	container := gabs.New()
-	_, _ = container.SetP(o.Username, "username")
-	_, _ = container.SetP(o.Password, "password")
 
-	body := container.String()
+	absCertPath, err := filepath.Abs(o.Cert)
+	if err == nil {
+		o.Cert = absCertPath
+	}
 
-	jsonParsed, err := util.EdgeControllerLogin(host, o.Cert, body, o.Out, o.OutputJSONResponse, o.edgeOptions.Timeout, o.edgeOptions.Verbose)
+	loginIdentity := &util.RestClientIdentity{
+		Url:       host,
+		Token:     o.Token,
+		LoginTime: time.Now().Format(time.RFC3339),
+		Cert:      o.Cert,
+		ReadOnly:  o.ReadOnly,
+	}
 
+	config, configFile, err := util.LoadRestClientConfig()
 	if err != nil {
 		return err
 	}
 
-	if !jsonParsed.ExistsP("data.token") {
-		return fmt.Errorf("no session token returned from login request to %v. Received: %v", host, jsonParsed.String())
+	id := config.GetIdentity()
+	fmt.Printf("Saving identity '%v' to %v\n", id, configFile)
+	config.Identities[id] = loginIdentity
+
+	err = util.PersistRestClientConfig(config)
+
+	return err
+}
+
+func (o *loginOptions) ConfigureCerts(host string, ctrlUrl *url.URL) error {
+	selfSigned, err := util.IsServerSelfSigned(host)
+	if err != nil {
+		return err
 	}
 
-	token, ok := jsonParsed.Path("data.token").Data().(string)
+	if selfSigned && o.Cert == "" {
+		serverCertChain, certs, err := util.GetServerCertChain(host)
+		if err != nil {
+			return errors.Wrapf(err, "unable to retrieve server certificate chain from %v", host)
+		}
+
+		savedCerts, certFile, err := util.ReadCert(ctrlUrl.Hostname())
+		if err != nil {
+			return err
+		}
 
-	if !ok {
-		return fmt.Errorf("session token returned from login request to %v is not in the expected format. Received: %v", host, jsonParsed.String())
+		if savedCerts != nil {
+			o.Cert = certFile
+			if !util.AreCertsSame(serverCertChain, savedCerts) {
+				fmt.Printf("WARNING: server supplied cert chain doesn't match cached certs at %v\n", certFile)
+				fmt.Printf("Issuer: %v\n", certs[0].Issuer.String())
+				replace, err := o.askYesNo("Replace cached certs [Y/N]: ")
+				if err != nil {
+					return err
+				}
+				if replace {
+					_, err = util.WriteCert(ctrlUrl.Hostname(), serverCertChain)
+					if err != nil {
+						return err
+					}
+				}
+			}
+		} else {
+			if len(certs) == 0 {
+				fmt.Print("WARNING: server supplied cert chain was empty, continuing but login will likely fail")
+			} else {
+				fmt.Printf("Self-signed server certificate chain retrieved from server\n")
```

Comment:
Most ziti instances aren't 'self-signed'. Self-signing is using a private key to sign its own certificate.

In ziti, we are usually dealing with an "untrusted CA". CAs are self-signed certs, but for clarity, they are called CAs because they have the ability to delegate trust via the CA flag in their certificate extensions.

I would change the wording to "Unknown certificate authority retrieved from sever".

Reply:
fixed